Lab 4

AntiFrizz1 opened the pull request.

MChepulis committed:
add env cubemap

add HDRI texture file and it`s convertation to cuvemap texture

add environment cubemap texture

FIXME: bad code

MChepulis committed:
add irradiance map convertation

add conversion of env_cubemap to irradiance map

MChepulis committed:
add ambient diffuse (irradiance)

add function for ambient diffuse (irradiance)
apdate Fresnel fuction
add several env texture
Fix hole-bug if env. sphere

MChepulis committed:
add depth stencil

add depth stencil initialization
add depth stencil resizing in resize-method
add depth test in RenderFrame

MChepulis committed:
update constant buffers

split ConstantBuffer to
- ConstantBuffer
- LightsBuffer
- MaterialBuffer

MChepulis committed:
Create Lab_4 project

rename Lab_3 project to Lab_4

MChepulis committed:
Revert "Create Lab_4 project"

This reverts commit d8866ef6854c13834203258b41242b888ec4567b.

MChepulis committed:
add lab_4 project

AntiFrizz1 committed:
Merge remote-tracking branch 'origin/main' into lab_4-from-lab_3

AntiFrizz1 committed:
set c++17

MChepulis committed:
replace compute of irradiance texture

replace compute of irradiance texture from RanderFrame to Initialize

anthark reviewed (changes requested):
В принципе, хорошо, но есть несколько проблем, которые нужно исправить: 
1) В G_SCHLICKGGX нужно ограничить dot(n, v) снизу нулем, поскольку если там получится мелкое отрицательное значение (в силу точности вычислений), в знаменателе может оказаться 0, со всеми вытекающими последствиями
2) Skybox сфера должна двигаться вместе с камерой, так чтобы ее центр всегда оставался в точке положения наблюдателя
3) Roughness не должен быть == 0, опять же проблемы с nan/inf возникнут
4) С настройками по умолчанию приложение не запускается из-под РендерДока

MChepulis committed:
fix remarks

1) limited dot(n, v) from below in G_SCHLICKGGX
2) fix the center of the env. sphere at the camera position
3) now roughness in [0.001, 1]
4) add copy texture files (.dds) to target dir during build
5) add lab_4 description in README

anthark reviewed (approved):
ОК, задача засчитана
